### DISCUSSION TITLE
VRC World Support

### DISCUSSION THREAD

**Discussion by mitsukaki** [bdunderscore/ndmf#242] [Ideas]
Cheers all,

I hope I'm posting in the appropriate place. I was hoping to start some discussion around the potential of extending the NDM framework to support VRChat worlds. Is the necessary ground work on VRChat's end present for such a system?

Assuming that it were:
- Is there interest in developing such?
- Are there any challenges you could foresee in expanding NDMF in this way?

Looking forward to hearing more perspectives :)

**Comment by bdunderscore** [comment]
There’s a hook system in VRCSDK for worlds as well - I’m not sure how necessary it is to have an orchestration system like NDMF as worlds are usually more modular from the start. A lot of the complexity leading to NDMF comes from the need to coordinate changes to the avatar descriptor and animator controllers.

Is there a specific use case you have in mind?

**Reply by mitsukaki** [reply]
> There’s a hook system in VRCSDK for worlds as well

For my own research I looked this up quickly. Just to be sure I'm looking at the right place, you were referring to the VRChat SDK API, right?

> A lot of the complexity leading to NDMF comes from the need to coordinate changes to the avatar descriptor and animator controllers.

This is a reasonable point-  if the core value of NDMF was to mediate between plugins contending for the avatar descriptor and such, I agree it makes less sense for a world edition of this, given we don't really have those contentions there.  

> Is there a specific use case you have in mind?

I hoped to be able to create some procedural tools- things that can generate animators, or do more involved optimizations. (As some examples, combining meshes, generating atlases, or a component that tried to resize textures in specified objects to fit in a texture budget).

I'll need to dig deeper into VRChat's SDK API- I wasn't previously aware of it- but maybe it already offers enough to pull off these things that I had in mind.

**Comment by bdunderscore** [comment]
I don’t have a reference handy right now, but take a look at how U# performs prebuild processing to see how to hook in.

in any case, I don’t currently have a clear vision for how to extend NDMF to worlds, but I’m not necessarily opposed to it if it makes sense. I do plan to extract some subsystems which could be useful in worlds however (some of which I’m prototyping in https://github.com/bdunderscore/ndmf-core right now).
